lib/main.dart

```diff
@@ -42,13 +42,31 @@ class SampleViewerApp extends StatefulWidget {
 }
 
 class _SampleViewerAppState extends State<SampleViewerApp> {
-  final _samples = <Sample>[];
+  final _allSamples = <Sample>[];
+  final _searchFocusNode = FocusNode();
+  final _textEditingController = TextEditingController();
+  var _filteredSamples = <Sample>[];
   bool _ready = false;
+  bool _searchHasFocus = false;
```

Comment:
Is this needed? Why do we need both `_searchHasFocus` and `_searchFocusNode` which has a `hasFocus` property?

Reply:
It's not strictly necessary, but the problem is that `_searchFocusNode` can change its `hasFocus` depending on user actions, and that doesn't in and of itself cause a rebuild of the widget tree. So we need to listen for changes to `_searchFocusNode` to detect when to call `setState()` (which causes the actual rebuild). That requires just a small bit of logic down below in `_searchFocusNode.addListener(...)`.

Reply:
In that case can we have `_searchFocusNode.addListener` call setState without needing to duplicate that property? Or, only do that somehow when hasFocus changed without needing to maintain a separate bool?

If not, no problem.

Reply:
The `addListener` call is over-inclusive. There are other changes that can trigger the `addListener` callback, but we only want to call `setState()` if `hasFocus` changes. As far as I know, there's no way to get a notification for only `hasFocus`. So we have to listen for everything, and from there check for a change specifically to `hasFocus`.

The other nice thing is that `setState()` is self-documenting, in that you can see what property is being changed. If you did just a generic empty `setState(() {})`, you kinda lose the thread of what's going on.